java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java

```diff
@@ -68,4 +71,45 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
         for (JsonObject composite : composites)
             pePlacementPreprocess.compositeColocateIdUse(composite);
     }
+
+    @SuppressWarnings("serial")
+    private void relocateHashAdders(){
+        final Set<JsonObject> hashAdders = new HashSet<>();
+        // Firstly, find each hashAdder
+        GraphUtilities.visitOnce(GraphUtilities.findStarts(graph),
+                new HashSet<BVirtualMarker>(),
+                graph,
+                new Consumer<JsonObject>(){
+                    public void accept(JsonObject op) {
+                        if(jstring(op, "kind").equals("com.ibm.streamsx.topology.functional.java::HashAdder")){
```

Comment:
This is insufficient, and one of the reasons the old code was removed. What is actually needed is a property of the operator invocation to mark it as a logical hash adder, and not determine it from its kind.

Reply:
Should I add that property to `OpPerperties` and `BOperatorInvocation#_json()` similar to what is done for `KIND`? Any suggestion for the property name? What about `HASHADDER_OP`?

Reply:
Boolean JSON property "hashAdder" added in OpProperties.

Reply:
Do you mean this [`OpProperties`](https://github.com/IBMStreams/streamsx.topology/blob/master/java/src/com/ibm/streamsx/topology/generator/operator/OpProperties.java)? I didn't see a `"hashAdder"` property.

Reply:
this is not addressed yet